Compare branches

HemanthMudalaiah opened the pull request.

Commit:
This commit will have changes in rubycritic where you can compare two branches

Commit:
This commit has changes where we can decide which of the file changes degraded the code quality

Commit:
Comparing the diff with the whole branch for duplication analysis

Commit:
Configuring smell link to point to feature branch code

Commit:
Changes in the content of the build details

Commit:
Gitlab Integration

Pushing the status of the code and report location to gitlab using gitlab api

Commit:
Code cleanup

* Code refactoring, DRY
* Comparing between branches in local without pr id.

Commit:
Specs and code cleanup

1. Setting a new mode to compare between branches.
2. Moved the code changes to a seperate file.
3. Pushing comments on to pr.
4. Specs for analysed_module_collection.

Commit:
Code Cleanup & Refactoring

1. Removed Gitlab,Github integrations.
2. Template Changes.
3. Options changes.

Commit:
Added specs and threshold option

1. Added specs for compare module
2. Added threshold as an option in rubycritic before it was as app side configuration
3. Check for any uncommited changes

Commit:
Refactoring compare module

1. Removed redundant code.
2. Adding config so that browser doesn't open html report during test.

Commit:
Code cleanup

1. Changes for stub in compare test file
2. Cleaning up compare.rb
3. Changes in option parsing

Commit:
Options and compare changes

Commit:
Fixing the options fetaure spec

Commit:
Code cleanup and changes

1. Generating report for all modified files in the feature branch instead of degraded files as before.
2. Removing methods related to get degraded files.
3. Fixing specs due the above changse.

Commit:
Spec fix

Commit:
Abort message changes

Commit:
Adding links in the navbar

Commit:
Build fixes

rohitcy committed:
[Fixes #187] Fix code navigation links

tejasbubane commented on features/command_line_interface/options.feature:
```diff
@@ -21,6 +21,10 @@ Feature: RubyCritic can be controlled using command-line options
           -p, --path [PATH]                Set path where report will be saved (tmp/rubycritic by default)
+          -b BASE_BRANCH,FEATURE_BRANCH,   Set branches to compare
+              --branch
+          -t [THRESHOLD_SCORE],            Set a threshold score works only with -b option
```
threshold for score difference between two branches (works only with -b)

tejasbubane commented on lib/rubycritic/commands/compare.rb:
```diff
@@ -0,0 +1,115 @@
+    class Compare < Default
+      def initialize(options)
+        super
+        @number = 0
```
rename to `@build_number`

tejasbubane commented on lib/rubycritic/generators/html/view_helpers.rb:
```diff
@@ -26,7 +26,16 @@ def file_path(file)
-      file_path("#{location.pathname.sub_ext('.html')}#L#{location.line}")
+      smell_location = "#{location.pathname.sub_ext('.html')}#L#{location.line}"
+      if Config.mode == :compare_branches
+        "file://#{File.expand_path(Config.feature_root_directory)}/#{smell_location}"
```
use `file_path`

tejasbubane commented on lib/rubycritic/generators/html/view_helpers.rb:
```diff
@@ -26,7 +26,16 @@ def file_path(file)
+    end
+
+    def code_index_path(root_directory)
+      "file://#{File.expand_path(root_directory)}/code_index.html"
```
use `file_path`

tejasbubane commented on lib/rubycritic/generators/html_report.rb:
```diff
@@ -64,7 +65,7 @@ def copy_assets_to_report_directory
 
       def report_location
-        overview_generator.file_href
+        Config.mode == :compare_branches ? code_index_generator.file_href : overview_generator.file_href
```
Config.compare_branches_mode?

Commit:
PR Changes

tejasbubane reviewed (approved):
@HemanthMudalaiah Ok dude, go ahead and submit a PR to the upstream repo.

rohitcy committed:
[Fixes #60] Implement search and filter on code and smells view

Commit:
Add missing method 'head_reference' for Perforce source control

tejasbubane committed:
[Fixes #189] Use Ruby's File instead of `wc` system command

As indicated by @guilhermesimoes
https://github.com/whitesmith/rubycritic/pull/184#discussion_r88904547

troessner committed:
Merge pull request #193 from tejasbubane/fix-usage-of-wc

[Fixes #189] Use Ruby's File instead of `wc` system command

nunosilva800 committed:
Merge pull request #190 from rohitcy/fix-code-navigation-links

[Fixes #187] Fix code navigation links

nunosilva800 committed:
Merge pull request #194 from Rataah/master

Add missing method 'head_reference' for Perforce source control

nunosilva800 committed:
Merge pull request #192 from rohitcy/search-filter-feature

[Fixes #60] Implement search and filter on code and smells view

nunosilva800 committed:
Bump to 3.1.1

troessner committed:
Merge pull request #195 from whitesmith/bump-to-3.1.1

Bump to 3.1.1

josephpage committed:
Add MRI 2.3.3 compatibility (update parser gem)

troessner committed:
Merge pull request #196 from josephpage/patch-1

Add MRI 2.3.3 compatibility

nunosilva800 committed:
Fix release notes

troessner committed:
Merge pull request #197 from whitesmith/fix-release-notes

Fix release notes

hoshinotsuyoshi committed:
Make sure that rating value is a String

tejasbubane committed:
Add ruby 2.3.3 to travis build

We upgraded parser to version 2.3.3 but we need to make sure the build
passes for that version.

troessner committed:
Merge pull request #199 from tejasbubane/add-ruby-2.3.3

Add ruby 2.3.3 to travis build

nunosilva800 committed:
Drop the minor version from travis

nunosilva800 committed:
Merge pull request #200 from whitesmith/update-travis

Drop the minor version from travis

Commit:
Review changes and improvements

1. Adding up or down arrow before the rating in code index page depending on the status of the file.
2. Review changes.

tejasbubane committed:
Fix errors when no source-control is found

Churn calculation requires source-control.
Do not show the churn-complexity graph when source control is not found.

troessner committed:
Merge pull request #203 from tejasbubane/handle-no-sourcecontrol

Fix errors when no source-control is found

nunosilva800 committed:
Merge pull request #198 from hoshinotsuyoshi/fix_not_displaying_gpa-chart_in_rails

Running "rake rubycritic" in rails project does not display gpa-chart(in RubyCritic version 3.1.0)

nunosilva800 committed:
Bump to v3.1.2

troessner committed:
Merge pull request #204 from whitesmith/bump-to-3.1.2

Bump to v3.1.2

tejasbubane committed:
Use markdown reference-style links for github usernames in CHANGELOG

Inspired by [Rubocop's CHANGELOG](https://github.com/bbatsov/rubocop/blob/master/CHANGELOG.md).

tejasbubane committed:
Add Changelog entry format to the Contributing guide

troessner committed:
Merge pull request #205 from tejasbubane/use-github-links-in-changelog

[ci skip] Use GitHub links in changelog

thedrow committed:
Turn light red to orange since light red is no longer available.

nunosilva800 committed:
Merge pull request #207 from thedrow/light-red-to-orange

Turn light red to orange since light red is no longer available

nunosilva800 committed:
Specify 'rainbow' version dependency

nunosilva800 committed:
Bump to 3.1.3

troessner committed:
Merge pull request #208 from whitesmith/bump-to-3.1.3

Bump to 3.1.3

Commit:
Modifying header links to stay on the same page even of the branch is changed

Commit:
Build fixes

Commit:
Resolving merge conflicts

Commit:
Change in the threshold check condition.

jdickey committed:
Update RuboCop to 0.47.1 from 0.42.0.

Obvious change to `rubycritic.gemspec`. This commit also changes the
`.rubocop.yml` file to disable a `Security/MarshalLoad` cop (introduced
in RuboCop 0.47.0) from flagging the `RubyCritic::Serializer.load`
method. Interestingly, wrapping that method in RuboCop disable/enable
comments for that cop simply flags a `Lint/UnneededDisable` violation.
Go figure. Adding the one file to a disable-Security/MarshalLoad entry
in `.rubocop.yml` solves the problem, even if it does feel like shooting
a fly with an RPG.

nunosilva800 committed:
Merge pull request #217 from jdickey/update-rubocop

Update RuboCop to 0.47.1 from 0.42.0.

Commit:
Adding this  https://github.com/whitesmith/rubycritic/pull/191/files#r92583493

Commit:
Adding proper message in help

Commit:
Merge branch 'master' of github.com:whitesmith/rubycritic into compare_branches
